tests/test_install.sh

```diff
@@ -1,14 +1,155 @@
 #! /bin/sh
 
-test_dir=$(dirname $0)
-mender_configure=${test_dir}/../src/mender-configure
+test_dir="$(dirname $0)"
+tmp_dir="$(mktemp -d)"
+state_dir="$(mktemp -d)"
+module_dir="${state_dir}/module"
+applycmd_dir="$(mktemp -d)"
+applycmd_log="${tmp_dir}/apply-device-config.log"
+mender_configure="${test_dir}/../src/mender-configure"
 
+# Override variables inside the Update Module.
+export TEST_CONFIG="${state_dir}/device-config.json"
+export TEST_APPLY_CMD="${applycmd_dir}/apply-device-config"
 
-testArtifactInstallDummy() {
-    output=$(./$mender_configure ArtifactInstall /tmp/nonexisting)
-    exit_code=$?
-    assertEquals 0 $exit_code
-    assertEquals "Dummy install" "${output}"
+oneTimeTearDown() {
+    rm -rf "${tmp_dir}"
+}
+
+setUp() {
+    mkdir -p "${tmp_dir}"
+    mkdir -p "${state_dir}"
+    mkdir -p "${applycmd_dir}"
+    mkdir -p "${module_dir}"
+
+    cat > "${TEST_APPLY_CMD}" <<EOF
+#!/bin/sh
+
+echo "\$(basename "\$0")" "\$@" >> "${applycmd_log}"
+EOF
+    chmod ugo+x "${TEST_APPLY_CMD}"
+
+    cat > "${module_dir}/meta-data" <<EOF
+{"new":"value"}
+EOF
+
+    echo '{"old":"value"}' > "${TEST_CONFIG}"
+}
+
+tearDown() {
+    rm -rf "${state_dir}" "${applycmd_dir}" "${tmp_dir}"
+}
+
+
+testOtherState() {
+    output="$(./$mender_configure Download "${module_dir}")"
+    assertEquals 0 $?
+    assertEquals "" "${output}"
+}
+
+testFreshArtifactInstall() {
+    rm -f "${TEST_CONFIG}"
+
+    output="$(./$mender_configure ArtifactInstall "${module_dir}")"
+    assertEquals 0 $?
+    assertEquals "" "${output}"
+
+    assertEquals '{"new":"value"}' "$(cat "${TEST_CONFIG}")"
+    assertEquals "apply-device-config ${TEST_CONFIG}" "$(cat "${applycmd_log}")"
+}
+
+testArtifactInstallConfigExists() {
+    output="$(./$mender_configure ArtifactInstall "${module_dir}")"
+    assertEquals 0 $?
+    assertEquals "" "${output}"
+
+    assertEquals '{"new":"value"}' "$(cat "${TEST_CONFIG}")"
+    assertEquals "apply-device-config ${TEST_CONFIG}" "$(cat "${applycmd_log}")"
+}
+
+testNeedsArtifactRebootAutomatic() {
+    echo 'exit 20' >> "${TEST_APPLY_CMD}"
+
+    output="$(./$mender_configure ArtifactInstall "${module_dir}")"
+    assertEquals 0 $?
+    assertEquals "" "${output}"
+
+    output="$(./$mender_configure NeedsArtifactReboot "${module_dir}")"
+    assertEquals 0 $?
+    assertEquals "Automatic" "${output}"
+
+    assertEquals '{"new":"value"}' "$(cat "${TEST_CONFIG}")"
+    assertEquals "apply-device-config ${TEST_CONFIG}" "$(cat "${applycmd_log}")"
+}
+
+testNeedsArtifactRebootNo() {
+    output="$(./$mender_configure ArtifactInstall "${module_dir}")"
+    assertEquals 0 $?
+    assertEquals "" "${output}"
+
+    output="$(./$mender_configure NeedsArtifactReboot "${module_dir}")"
+    assertEquals 0 $?
+    assertEquals "No" "${output}"
+
+    assertEquals '{"new":"value"}' "$(cat "${TEST_CONFIG}")"
+    assertEquals "apply-device-config ${TEST_CONFIG}" "$(cat "${applycmd_log}")"
+}
+
```

Comment:
Mmmm, what about a test case calling `NeedsArtifactReboot` without a previous call to `ArtifactInstall`? This is not now the client will use the module, but it should still be robust and reply "No".

Maybe testing this is a bad idea, actually, but I just wanted to ask out loud. I think it might be a bad idea because to support such thing we should have an API to call the configure script (with no arguments, for example) and it shall return 20/0 depending on the default preference.

Thoughts?

Reply:
> Mmmm, what about a test case calling `NeedsArtifactReboot` without a previous call to `ArtifactInstall`? This is not now the client will use the module, but it should still be robust and reply "No".

Agree, I will add this.

> Maybe testing this is a bad idea, actually, but I just wanted to ask out loud. I think it might be a bad idea because to support such thing we should have an API to call the configure script (with no arguments, for example) and it shall return 20/0 depending on the default preference.

I didn't quite understand this. What API do you mean?

Reply:
> I didn't quite understand this. What API do you mean?

I don't mean "API" in the strict sense, but in the idea of the agreement the module has with the configure script or tool.

What I was thinking is that the one possible valid use case of calling `NeedsArtifactReboot` before `ArtifactInstall` is if we could call the configure script (for example, with no arguments) to learn about if the script requires reboot or not. Now, we don't have such agreement (abstract API) with the script, so the use case does fall apart completely.

Regardless of being a valid use case or not, for the sake of robustness, we should test it (as you have added). So far so good :)

Reply:
Right, ok. You're right, semantically it makes sense, but the client will never call it in this order (*). In any case, we are robust enough to handle it if it ever happens. :-)

(*) This is a deliberate design choice: If `ArtifactInstall` contains `apt install`, then you won't know whether you have to reboot until afterwards, so this is an explicit promise in the API, that they will be called in that order.